File: prd-admin/src/pages/product-agent/tapdDefectFieldCatalog.ts
/**
 * TAPD 缺陷字段 SSOT — 与 prd-api TapdDefectFieldCatalog / CapsuleExecutor.MapBugFieldsToChinese 对齐。
 * 产品管理「严重程度」仅存 structuredData，不跨字段推断。
 */
import type { DefectSeverityLevel } from './defectSeverity';

export const TAPD_DEFECT_FIELD = {
  defectId: '缺陷ID',
  title: '标题',
  reporter: '创建人',
  created: '创建时间',
  issueStartTime: '问题开始时间',
  resolved: '解决时间',
  closed: '关闭时间',
  due: '预计结束时间',
  currentOwner: '处理人',
  status: '状态',
  responsiblePerson: '责任人',
  overdue: '是否逾期',
  validReport: '有效报告',
  defectGrade: '缺陷等级',
  /** V2.6 四档：致命/严重/一般/轻微 */
  defectSeverity: '严重程度',
  /** TAPD 导出列原文：紧急/高/中/低/无关紧要 */
  tapdSeveritySource: 'TAPD严重程度',
  defectDivision: '缺陷划分',
  feedbackPerson: '反馈人',
  companyName: '公司名称',
  merchantNo: '商户编号',
  introducedProject: '引入项目',
  feedbackTime: '反馈时间',
  impactScope: '影响范围',
  structureParent: '结构归母',
  logicAttribution: '逻辑归因',
  urlLink: 'URL链接',
  linksInDescription: '描述中的链接',
  isHistorical: '是否历史问题',
  timelyFixed: '及时处理',
} as const;

/** TAPD API custom_field 映射（文档用，导入器消费） */
export const TAPD_API_FIELD_MAP: Record<string, string> = {
  custom_field_100: TAPD_DEFECT_FIELD.issueStartTime,
  custom_field_two: TAPD_DEFECT_FIELD.responsiblePerson,
  custom_field_four: TAPD_DEFECT_FIELD.overdue,
  custom_field_five: TAPD_DEFECT_FIELD.validReport,
  custom_field_6: TAPD_DEFECT_FIELD.defectGrade,
  custom_field_7: TAPD_DEFECT_FIELD.defectDivision,
  custom_field_8: TAPD_DEFECT_FIELD.feedbackPerson,
  custom_field_9: TAPD_DEFECT_FIELD.companyName,
  custom_field_10: TAPD_DEFECT_FIELD.merchantNo,
  custom_field_11: TAPD_DEFECT_FIELD.introducedProject,
  custom_field_12: TAPD_DEFECT_FIELD.feedbackTime,
  custom_field_13: TAPD_DEFECT_FIELD.impactScope,
  custom_field_one: TAPD_DEFECT_FIELD.structureParent,
  custom_field_three: TAPD_DEFECT_FIELD.logicAttribution,
};

export type TapdFieldEditorKind = 'text' | 'textarea' | 'date' | 'readonly' | 'url' | 'severity' | 'classification';

export interface TapdDefectSidebarField {
  key: string;
  label: string;
  kind: TapdFieldEditorKind;
  entitySource?: 'externalId' | 'status' | 'reporterName' | 'createdAt' | 'resolvedAt' | 'closedAt' | 'assigneeName';
}

export const TAPD_DEFECT_SIDEBAR_FIELDS: TapdDefectSidebarField[] = [
  { key: TAPD_DEFECT_FIELD.defectId, label: 'ID', kind: 'readonly', entitySource: 'externalId' },
  { key: TAPD_DEFECT_FIELD.status, label: '状态', kind: 'readonly', entitySource: 'status' },
  { key: TAPD_DEFECT_FIELD.currentOwner, label: '处理人', kind: 'text' },
  { key: TAPD_DEFECT_FIELD.reporter, label: '创建人', kind: 'readonly', entitySource: 'reporterName' },
  { key: TAPD_DEFECT_FIELD.created, label: '创建时间', kind: 'readonly', entitySource: 'createdAt' },
  { key: TAPD_DEFECT_FIELD.defectSeverity, label: '严重程度', kind: 'severity' },
  { key: TAPD_DEFECT_FIELD.defectDivision, label: '缺陷划分', kind: 'classification' },
  { key: TAPD_DEFECT_FIELD.responsiblePerson, label: '责任人', kind: 'text' },
  { key: TAPD_DEFECT_FIELD.overdue, label: '是否逾期', kind: 'text' },
  { key: TAPD_DEFECT_FIELD.validReport, label: '有效报告', kind: 'text' },
  { key: TAPD_DEFECT_FIELD.feedbackPerson, label: '反馈人', kind: 'text' },
  { key: TAPD_DEFECT_FIELD.companyName, label: '公司名称', kind: 'text' },
  { key: TAPD_DEFECT_FIELD.merchantNo, label: '商户编号', kind: 'text' },
  { key: TAPD_DEFECT_FIELD.introducedProject, label: '引入项目', kind: 'text' },
  { key: TAPD_DEFECT_FIELD.feedbackTime, label: '反馈时间', kind: 'date' },
  { key: TAPD_DEFECT_FIELD.impactScope, label: '影响范围', kind: 'textarea' },
  { key: TAPD_DEFECT_FIELD.structureParent, label: '结构归母', kind: 'text' },
  { key: TAPD_DEFECT_FIELD.logicAttribution, label: '逻辑归因', kind: 'text' },
  { key: TAPD_DEFECT_FIELD.issueStartTime, label: '问题开始时间', kind: 'date' },
  { key: TAPD_DEFECT_FIELD.due, label: '预计结束时间', kind: 'date' },
  { key: TAPD_DEFECT_FIELD.resolved, label: '解决时间', kind: 'readonly', entitySource: 'resolvedAt' },
  { key: TAPD_DEFECT_FIELD.closed, label: '关闭时间', kind: 'readonly', entitySource: 'closedAt' },
  { key: TAPD_DEFECT_FIELD.urlLink, label: 'URL链接', kind: 'url' },
  { key: TAPD_DEFECT_FIELD.isHistorical, label: '是否历史问题', kind: 'readonly' },
  { key: TAPD_DEFECT_FIELD.timelyFixed, label: '及时处理', kind: 'readonly' },
];

/** 仅根据已有日期字段派生只读标记；无足够数据时不写入默认值 */
export function computeTapdDerivedFields(structured: Record<string, string>): Record<string, string> {
  const issueStart = structured[TAPD_DEFECT_FIELD.issueStartTime]?.trim();
  const resolved = structured[TAPD_DEFECT_FIELD.resolved]?.trim();
  const due = structured[TAPD_DEFECT_FIELD.due]?.trim();
  const next = { ...structured };
  if (issueStart && resolved) {
    const startDt = new Date(issueStart);
    const resolvedDt = new Date(resolved);
    if (!Number.isNaN(startDt.getTime()) && !Number.isNaN(resolvedDt.getTime())) {
      const months = (resolvedDt.getFullYear() - startDt.getFullYear()) * 12 + (resolvedDt.getMonth() - startDt.getMonth());
      next[TAPD_DEFECT_FIELD.isHistorical] = months >= 6 ? '是' : '否';
    }
  }
  if (due && resolved) {
    const dueDt = new Date(due);
    const resolvedDt = new Date(resolved);
    if (!Number.isNaN(dueDt.getTime()) && !Number.isNaN(resolvedDt.getTime())) {
      next[TAPD_DEFECT_FIELD.timelyFixed] = dueDt.toDateString() >= resolvedDt.toDateString() ? '是' : '否';
    }
  }
  return next;
}

export function tierToStructuredValue(level: DefectSeverityLevel | ''): string {
  return level;
}

File: prd-api/src/PrdAgent.Core/Models/TapdDefectFieldCatalog.cs
namespace PrdAgent.Core.Models;

/// <summary>
/// TAPD 缺陷字段 SSOT（与 CapsuleExecutor.MapBugFieldsToChinese / TAPD API custom_field_* 对齐）。
/// 导入与产品内详情页共用同一套中文 key，值落在 DefectReport.StructuredData 或实体字段。
/// </summary>
public static class TapdDefectFieldCatalog
{
    public const string DefectId = "缺陷ID";
    public const string Title = "标题";
    public const string Reporter = "创建人";
    public const string Created = "创建时间";
    public const string IssueStartTime = "问题开始时间";
    public const string Resolved = "解决时间";
    public const string Closed = "关闭时间";
    public const string Due = "预计结束时间";
    public const string CurrentOwner = "处理人";
    public const string Status = "状态";
    public const string ResponsiblePerson = "责任人";
    public const string Overdue = "是否逾期";
    public const string ValidReport = "有效报告";
    /// <summary>TAPD 自定义「缺陷等级」(custom_field_6)，与标准导出列「严重程度」不同。</summary>
    public const string DefectGrade = "缺陷等级";
    /// <summary>产品管理 V2.6 严重程度（致命/严重/一般/轻微），存 StructuredData。</summary>
    public const string DefectSeverity = "严重程度";
    /// <summary>TAPD 导出「严重程度」列原文（紧急/高/中/低/无关紧要），导入镜像。</summary>
    public const string TapdSeveritySource = "TAPD严重程度";
    public const string DefectDivision = "缺陷划分";
    public const string FeedbackPerson = "反馈人";
    public const string CompanyName = "公司名称";
    public const string MerchantNo = "商户编号";
    public const string IntroducedProject = "引入项目";
    public const string FeedbackTime = "反馈时间";
    public const string ImpactScope = "影响范围";
    public const string StructureParent = "结构归母";
    public const string LogicAttribution = "逻辑归因";
    public const string UrlLink = "URL链接";
    public const string LinksInDescription = "描述中的链接";
    public const string IsHistorical = "是否历史问题";
    public const string TimelyFixed = "及时处理";

    /// <summary>详情页右侧属性栏展示顺序（与 TAPD 查看页一致）。</summary>
    public static readonly string[] SidebarFieldKeys =
    {
        DefectId, Status, CurrentOwner, Reporter, Created, DefectSeverity, DefectDivision,
        ResponsiblePerson, Overdue, ValidReport, FeedbackPerson, CompanyName, MerchantNo,
        IntroducedProject, FeedbackTime, ImpactScope, StructureParent, LogicAttribution,
        IssueStartTime, Due, Resolved, Closed, UrlLink, IsHistorical, TimelyFixed,
    };

    /// <summary>写入 StructuredData 时同步缺陷划分到 ProductDefectClassification。</summary>
    public static void SyncClassificationToEntity(DefectReport defect)
    {
        if (defect.StructuredData.TryGetValue(DefectDivision, out var div) && !string.IsNullOrWhiteSpace(div))
            defect.ProductDefectClassification = ProductDefectLinkageCatalog.NormalizeClassification(div);
        else if (!string.IsNullOrWhiteSpace(defect.ProductDefectClassification)
                 && !defect.StructuredData.ContainsKey(DefectDivision))
            defect.StructuredData[DefectDivision] = defect.ProductDefectClassification;
    }

    public static Dictionary<string, string> MergeStructuredData(
        Dictionary<string, string>? existing,
        Dictionary<string, string>? patch)
    {
        var merged = existing != null ? new Dictionary<string, string>(existing) : new Dictionary<string, string>();
        if (patch == null) return merged;
        foreach (var (k, v) in patch)
            merged[k] = v ?? string.Empty;
        return merged;
    }
}
